---
sidebar_position: 1
---

# Analysing Sanger sequence data

Welcome!  In this tutorial we will conduct a basic analysis of some Sanger sequence data. The data comes from a PCR of
two regions around the *HBA1* and *HBA2* genes that encode alpha-globin in three samples. This practical will walk you
through basic steps in analysing them.

## Prerequisites

To get started you need to download your group's data, and you'll also need some software to view the trace files and
extract the sequence.

### Getting the software

To view Sanger trace files, we'll use a software package called **FinchTV**. Start by downloading it from [the FinchTV
website](https://digitalworldbiology.com/FinchTV) (click on 'Download FinchTV for Windows' or 'Download FinchTV for MacOS'). 

After downloading, open the zip file and double-click on the `FinchTV_<version>` directory - you can then install
FinchTV by running the installer.

:::tip Note

In case of problems, a backup copy can also be found
[here](https://www.well.ox.ac.uk/bioinformatics/training/MSc_GM_2022/08_alpha-globin_experiment_analysis/software/). Another
similar program is [Chromas](http://technelysium.com.au/wp/chromas/) which can also be found in that folder.

:::

### Getting the data

Some sanger sequencing data from this experiment can be found in [this
folder](https://www.well.ox.ac.uk/bioinformatics/training/MSc_GM_2022/08_alpha-globin_experiment_analysis/data/attempt1/).

The data is separated into 'groups', but all groups ran the same experiment.  For the purposes of this tutorial, pick a
group and download the corresponding zip file by clicking in your web browser. You should now be able to open the zip file -
drag the resulting folder called `HBA_practical_group_[n]` to your desktop.

You're ready to start - now go and [view the files](introduction.md).